New fields in Segment.php

mikee2 opened the pull request:
New fields in macro replacement function

Add additional needed fields so that you can invoice yearly items. Have number and text descriptions to avoid future changes that may break existing invoices.

mikee2 committed:
Update Segment.php

Add additional needed fields so that you can invoice yearly items. Have number and text descriptions to avoid future changes that may break existing invoices.

eldy commented on htdocs/includes/odtphp/Segment.php:
```diff
@@ -170,8 +170,16 @@ public function macroReplace($text)
 
-        $patterns=array( '__CURRENTDAY__','__CURRENTDAYTEXT__','__CURRENTMONTHSHORT__','__CURRENTMONTH__','__CURRENTYEAR__' );
-        $values=array( date('j'), $langs->trans(date('l')), $langs->trans(date('M')), $langs->trans(date('F')), date('Y') );
+        $nextMonth = strtotime('+1 month');
```
using strtotime does not take in consideration setup of user and timezone.
Can you use instead dolibarr function like dol_get_next_month, dol_getdate, dol_mktime, ...

mikee2 committed:
Updated to use dolibarr timme functions

Redefined macro strings to adapt to dolibarr date functions return values. Keep same macro names when possible to avoid breaking existing code. Some macro holders (not commonly used) have been dropped because non existing values in dolibarr date functions.

Per eldy request changed php date functions with dolibarr's internal date functions to match users time zone. Now dates are those of the user's timezone so check that is is correctly defined in user's profile.

mikee2 commented:
Using dolibarr time functions cause a fatal error in propals. The funtion dol_get_month dos not exist in the propal definition chain so the changes to segment.php break the propal odt document generation:

[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] **PHP Fatal error:  Call to undefined function dol_get_next_month() in /var/www/dolibarr/htdocs/includes/odtphp/Segment.php on line 176**, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP Stack trace:, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   1. {main}() /var/www/dolibarr/htdocs/comm/propal/card.php:0, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   2. include() /var/www/dolibarr/htdocs/comm/propal/card.php:1224, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   3. Propal->generateDocument() /var/www/dolibarr/htdocs/core/actions_builddoc.inc.php:85, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   4. CommonObject->commonGenerateDocument() /var/www/dolibarr/htdocs/comm/propal/class/propal.class.php:3452, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   5. doc_generic_proposal_odt->write_file() /var/www/dolibarr/htdocs/core/class/commonobject.class.php:3973, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   6. Segment->merge() /var/www/dolibarr/htdocs/core/modules/propale/doc/doc_generic_proposal_odt.modules.php:445, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17
[Sun Jul 09 09:18:44 2017] [error] [client 192.168.10.55] PHP   7. Segment->macroReplace() /var/www/dolibarr/htdocs/includes/odtphp/Segment.php:143, referer: http://192.168.0.2/dolibarr/htdocs/comm/propal/card.php?id=17

mikee2 commented:
As a temporary solution for my own use I have added this at the top of the segment.php file:

if ( !function_exists ('dol_get_next_month') ) {
  include_once(DOL_DOCUMENT_ROOT.'/core/lib/date.lib.php');
}

But the preferred fix should be to make the function chain consistent for all modules calling/using, either directly or non directly, segment.php

eldy commented:
Version 6 introduced a more generic way to make substitution:

I recommend to use this part of code instead

        include_once DOL_DOCUMENT_ROOT.'/core/lib/date.lib.php';
    
        $substitutionarray=array();
            $tmp=dol_getdate(dol_now(), true);
            $tmp2=dol_get_prev_day($tmp['mday'], $tmp['mon'], $tmp['year']);
            $tmp3=dol_get_prev_month($tmp['mday'], $tmp['mon'], $tmp['year']);
            $tmp4=dol_get_next_day($tmp['mday'], $tmp['mon'], $tmp['year']);
            $tmp5=dol_get_next_month($tmp['mday'], $tmp['mon'], $tmp['year']);
        $substitutionarray=array_merge($substitutionarray, array(
            '__DAY__' => $tmp['mday'],
            '__MONTH__' => $tmp['mon'],
            '__YEAR__' => $tmp['year'],
            '__PREVIOUS_DAY__' => $tmp2['day'],
            '__PREVIOUS_MONTH__' => $tmp3['month'],
            '__PREVIOUS_YEAR__' => ($tmp['year'] - 1),
            '__NEXT_DAY__' => $tmp4['day'],
            '__NEXT_MONTH__' => $tmp5['month'],
            '__NEXT_YEAR__' => ($tmp['year'] + 1),
        ));

       complete_substitutions_array($substitutionarray, $langs);
       $text=make_substitutions($text,$substit);

Can you check with this and make a PR with that if it works like you expect ?

mikee2 commented:
Hi Laurent.

 

Do you mean that version 6 is out for us to use?  I am in version 5.0.3.

 

As there is a major version change, is there a change log that we can check?

 

Regards,

 

Miguel Erill

Doymer Consultores

 

 

De: Laurent Destailleur [mailto:notifications@github.com] 
Enviado el: martes, 18 de julio de 2017 15:09
Para: Dolibarr/dolibarr
CC: mikee2; Author
Asunto: Re: [Dolibarr/dolibarr] New fields in Segment.php (#6806)

 

Version 6 introduced a more generic way to make substitution:

I recommend to use this part of code instead

    include_once DOL_DOCUMENT_ROOT.'/core/lib/date.lib.php';
 
    $substitutionarray=array();
        $tmp=dol_getdate(dol_now(), true);
        $tmp2=dol_get_prev_day($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp3=dol_get_prev_month($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp4=dol_get_next_day($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp5=dol_get_next_month($tmp['mday'], $tmp['mon'], $tmp['year']);
    $substitutionarray=array_merge($substitutionarray, array(
        '__DAY__' => $tmp['mday'],
        '__MONTH__' => $tmp['mon'],
        '__YEAR__' => $tmp['year'],
        '__PREVIOUS_DAY__' => $tmp2['day'],
        '__PREVIOUS_MONTH__' => $tmp3['month'],
        '__PREVIOUS_YEAR__' => ($tmp['year'] - 1),
        '__NEXT_DAY__' => $tmp4['day'],
        '__NEXT_MONTH__' => $tmp5['month'],
        '__NEXT_YEAR__' => ($tmp['year'] + 1),
    ));
 
   complete_substitutions_array($substitutionarray, $langs);
   $text=make_substitutions($text,$substit);

Can you check with this and make a PR with that if it works like you expect ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/Dolibarr/dolibarr/pull/6806#issuecomment-316058191> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AFi3ZvVwD40m5fUb-omnPp3Eqf8kZaC5ks5sPK5pgaJpZM4NSr0D> .  <https://github.com/notifications/beacon/AFi3ZmCq6UGvqU1ZZqhwrlvVekjPp1ajks5sPK5pgaJpZM4NSr0D.gif>

braito4 commented:
Still in beta, but you can try It on a local machine with a copy of your bd

braito4 commented:
https://github.com/Dolibarr/dolibarr/blob/6.0/ChangeLog

mikee2 commented:
Hi Laurent.

 

Final code of the  macroReplace function for version 6 of dolibarr follows:

 

   /**

    * Function to replace macros for invoice short and long month, invoice year

    * 

    * Substitution occur when the invoice is generated, not considering the invoice date

    * so do not (re)generate in a different date than the one that the invoice belongs to

    * Perhaps it would be better to use the invoice issued date but I still do not know

    * how to get it here

    *

    * Miguel Erill 19/09/2017

    * 

    * @param       string          $value        String to convert

    */

    public function macroReplace($text)

    {

        global $langs;

 

//        $substitutionarray=array();

        $tmp=dol_getdate(dol_now(), true);

        $tmp2=dol_get_prev_day($tmp['mday'], $tmp['mon'], $tmp['year']);

        $tmp3=dol_get_prev_month($tmp['mon'], $tmp['year']);

        $tmp4=dol_get_next_day($tmp['mday'], $tmp['mon'], $tmp['year']);

        $tmp5=dol_get_next_month($tmp['mon'], $tmp['year']);

 

//        $substitutionarray=array_merge($substitutionarray, array(

        $substitutionarray=array(

          '__DAY__' => $tmp['mday'],

          '__CURRENTDAY__' => $tmp['mday'],

          '__MONTH__' => $tmp['mon'],

          '__CURRENTMONTH__' => $tmp['mon'],

//          '__MONTHTEXT__' => $tmp['month'],  <------Can’t use this as it deliver English text that is NOT translated afterwards

          '__MONTHTEXT__' => monthArray($langs)[$tmp['mon']],  <------ Should not be needed to use the months array directly here (see later)

          '__CURRENTMONTHLONG__' => monthArray($langs)[$tmp['mon']],

          '__YEAR__' => $tmp['year'],

          '__CURRENTYEAR__' => $tmp['year'],

          '__PREVIOUS_DAY__' => $tmp2['day'],

          '__PREVIOUS_MONTH__' => $tmp3['month'],

          '__PREVIOUS_YEAR__' => ($tmp['year'] - 1),

          '__NEXT_DAY__' => $tmp4['day'],

          '__NEXT_MONTH__' => $tmp5['month'],

          '__NEXTMONTH__' => $tmp5['month'],

          '__NEXTMONTHLONG__' => monthArray($langs)[$tmp5['month']],

          '__NEXT_YEAR__' => ($tmp['year'] + 1),

          '__NEXTYEAR__' => ($tmp['year'] + 1),

        );

 

        complete_substitutions_array($substitutionarray, $langs);

        $text=make_substitutions($text,$substitutionarray);

 

        return $text;

    }

 

Anyway there are some considerations to make regarding the source code you sent me and the changes needed to make it work. 

 

First I have duplicated some tags to keep backwards compatibility as it looks like there are other people using this from some other messages I have received. They should be not advertised anywhere if we want to deprecate them for the future and keep this function as global and small as possible.

 

Then there must be wrong code where you extracted the proposed initial code that may be worth a check (see your initial proposal later in this email if you are so kind).

 

        $substitutionarray=array();

I….

       $substitutionarray=array_merge($substitutionarray, array(….));

 

Defining an empty array and then calling the array_merge function to merge it with another array is not a very efficient method to build an array. Why not to define the array directly?:

 

        $substitutionarray=array(….);   <- Enough to fully define the array

 

May be there are some good reason to do this but I am not able to find it. Please feel free to tell me that I am wrong but, in the meantime, I have used this other way.

 

Then you use dol_get_prev_month and dol_get_next_month with THREE (3) parameters where those functions only accept two. The correct calls should be:

 

        $tmp3=dol_get_prev_month($tmp['mon'], $tmp['year']);

        $tmp5=dol_get_next_month($tmp['mon'], $tmp['year']);

 

But your code includes an initial, additional, $tmp['mday'] that is wrong, delivering a wrong result where used. That part of the code must be delivering wrong results.

 

Finally you call the replacement function with a wrong array name, invalidating all the previous work:

 

$text=make_substitutions($text,$substit);

 

$substit does not exist anywhere and should be $substitutionarray instead (if we want to use that long name). I like shorter variable names and $substArr matches a lot more my feelings but I accept the long name if you prefer it. But we must use the same name all along the function. One or the other, do not mind, but the same one.

 

There are also some weird things there. The use of monthArray should NOT be needed as ‘complete_substitutions_array’ should have converted all values to the actual language (why to send it the $langs variable if not? or why to call it in the first place if not doing any substitution at all?), but it does not do that. In fact it does nothing for us regarding that language task. So if I use $tmp['month'] then I get ‘September’ as the value (not the capital initial letter) and that value does NOT get converted to the destination language of the user as I expected. So I had to build the month string by using the same method in V5. Anyway this time the translated value I receive is ‘septiembre’ (note the lower case of the initial letter) when in the monthArray the original text is capitalized too. Weird.

 

Another problem that  am having is not related to dolibarr itself but it avoids me to get the desired result. I have installed dolibarr in a Centos 7 with Centos Web Panel system. Centos 7 does not provide neither LibreOffice nor JODConverter by default. Installing LibreOffice is not an issue but I do not know where to get JODConverter from. Can you help here, please?

 

I have found a link to download jodconverter-core-3.0-beta-4 but I do not know if this is enough as it does not have any installable code, only jar files. Also there is no java installed in the server so I’ll may need to install it but there exist java 1.6, 1.7 and 1.8. What is the recommended one?

 

Hope that now we can, finally, close this long standing issue that I was not able to work on before as I did not have a suitable platform to try. Thanks for your time and help.

 

Regards,

 

Miguel Erill

Doymer Consultores

 

 

De: Laurent Destailleur [mailto:notifications@github.com] 
Enviado el: martes, 18 de julio de 2017 15:09
Para: Dolibarr/dolibarr
CC: mikee2; Author
Asunto: Re: [Dolibarr/dolibarr] New fields in Segment.php (#6806)

 

Version 6 introduced a more generic way to make substitution:

I recommend to use this part of code instead

    include_once DOL_DOCUMENT_ROOT.'/core/lib/date.lib.php';
 
    $substitutionarray=array();
        $tmp=dol_getdate(dol_now(), true);
        $tmp2=dol_get_prev_day($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp3=dol_get_prev_month($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp4=dol_get_next_day($tmp['mday'], $tmp['mon'], $tmp['year']);
        $tmp5=dol_get_next_month($tmp['mday'], $tmp['mon'], $tmp['year']);
    $substitutionarray=array_merge($substitutionarray, array(
        '__DAY__' => $tmp['mday'],
        '__MONTH__' => $tmp['mon'],
        '__YEAR__' => $tmp['year'],
        '__PREVIOUS_DAY__' => $tmp2['day'],
        '__PREVIOUS_MONTH__' => $tmp3['month'],
        '__PREVIOUS_YEAR__' => ($tmp['year'] - 1),
        '__NEXT_DAY__' => $tmp4['day'],
        '__NEXT_MONTH__' => $tmp5['month'],
        '__NEXT_YEAR__' => ($tmp['year'] + 1),
    ));
 
   complete_substitutions_array($substitutionarray, $langs);
   $text=make_substitutions($text,$substit);

Can you check with this and make a PR with that if it works like you expect ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/Dolibarr/dolibarr/pull/6806#issuecomment-316058191> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AFi3ZvVwD40m5fUb-omnPp3Eqf8kZaC5ks5sPK5pgaJpZM4NSr0D> .  <https://github.com/notifications/beacon/AFi3ZmCq6UGvqU1ZZqhwrlvVekjPp1ajks5sPK5pgaJpZM4NSr0D.gif>

eldy committed:
Merge branch 'develop' into develop

eldy commented:
I replace $langs->trans with $langs->transnoentitiesnoconv  (i suppose we don't want html entities into text) ?
Is it ok for you ?

mikee2 commented:
Hi Laurent.

 

‘$langs->trans’ does not exist in the V6 code. If you refer to V5 code then I do not get the point; Can you detail what you are trying to achieve?

 

On the other side I agree that it will be very rare the need for html entities in the macro variables. But one never knows…

 

Regards,

 

Miguel Erill

Doymer Consultores

 

 

De: Laurent Destailleur [mailto:notifications@github.com] 
Enviado el: jueves, 21 de septiembre de 2017 11:42
Para: Dolibarr/dolibarr
CC: mikee2; Author
Asunto: Re: [Dolibarr/dolibarr] New fields in Segment.php (#6806)

 

I replace $langs->trans with $langs->transnoentitiesnoconv (i suppose we don't want html entities into text) ?
Is it ok for you ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/Dolibarr/dolibarr/pull/6806#issuecomment-331106693> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AFi3ZnRJlQd8Mugkk4M7s-pGSnfyOwCKks5ski-FgaJpZM4NSr0D> .  <https://github.com/notifications/beacon/AFi3ZpC4lr89jO9HHaca2H6nhFh2oPeoks5ski-FgaJpZM4NSr0D.gif>

mikee2 commented:
Hi Eldy.

Why is this still marked as PR to fix? Can you tell me where the problem is? This is solved long time ago.

Thanks.
